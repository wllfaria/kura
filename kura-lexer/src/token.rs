#[allow(clippy::module_inception)]
mod token;

mod kind;
mod operator;
mod primitive;
mod value;

pub use kind::Kind;
pub use operator::Operator;
pub use primitive::{FloatSizes, IntSizes, Primitive, UIntSizes};
pub use token::{IntoToken, Location, Token};
pub use value::Value;
