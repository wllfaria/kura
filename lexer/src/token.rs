#[derive(Debug)]
pub enum NumSize {
    F8,
    F16,
    F32,
    F64,
    I8,
    I16,
    I32,
    I64,
    U8,
    U16,
    U32,
    U64,
    Usize,
    Isize,
}

impl std::fmt::Display for NumSize {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            NumSize::F8 => write!(f, "f8"),
            NumSize::F16 => write!(f, "f16"),
            NumSize::F32 => write!(f, "f32"),
            NumSize::F64 => write!(f, "f64"),
            NumSize::I8 => write!(f, "i8"),
            NumSize::I16 => write!(f, "i16"),
            NumSize::I32 => write!(f, "i32"),
            NumSize::I64 => write!(f, "i64"),
            NumSize::U8 => write!(f, "u8"),
            NumSize::U16 => write!(f, "u16"),
            NumSize::U32 => write!(f, "u32"),
            NumSize::U64 => write!(f, "u64"),
            NumSize::Usize => write!(f, "usize"),
            NumSize::Isize => write!(f, "isize"),
        }
    }
}

impl TryFrom<&str> for NumSize {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(NumSize::I8),
            "i16" => Ok(NumSize::I16),
            "i32" => Ok(NumSize::I32),
            "i64" => Ok(NumSize::I64),
            "u8" => Ok(NumSize::U8),
            "u16" => Ok(NumSize::U16),
            "u32" => Ok(NumSize::U32),
            "u64" => Ok(NumSize::U64),
            "f8" => Ok(NumSize::F8),
            "f16" => Ok(NumSize::F16),
            "f32" => Ok(NumSize::F32),
            "f64" => Ok(NumSize::F64),
            "usize" => Ok(NumSize::Usize),
            "isize" => Ok(NumSize::Isize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Token<'tok> {
    Integer { value: u64, size: Option<NumSize> },
    SignedInteger { value: i64, size: Option<NumSize> },
    Float { value: f64, size: Option<NumSize> },
    Minus,
    String(&'tok str),
}

impl std::fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Integer { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::SignedInteger { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Minus => write!(f, "-"),
            Token::String(s) => write!(f, "{s}"),
        }
    }
}
