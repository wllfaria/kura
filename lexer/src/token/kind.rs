use miette::Error;

use super::{
    operator::Operator,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug)]
pub enum Kind<'tok> {
    Value(Value<'tok>),
    Op(Operator),

    Var,
    Const,
    Match,
    If,
    Else,
    Fun,
    Struct,
    Enum,
    Return,
    Eof,
}

impl<'tok> Kind<'tok> {
    pub fn infix_precedence(&self) -> Result<(u8, u8), Error> {
        match self {
            Kind::Op(Operator::Plus) | Kind::Op(Operator::Minus) => Ok((1, 2)),
            Kind::Op(Operator::Less)
            | Kind::Op(Operator::LessEqual)
            | Kind::Op(Operator::Greater)
            | Kind::Op(Operator::GreaterEqual) => Ok((3, 4)),
            Kind::Op(Operator::Star) | Kind::Op(Operator::Slash) => Ok((5, 6)),
            _ => miette::bail!("invalid operator with no precedence"),
        }
    }

    pub fn is_binary_op(&self) -> bool {
        matches!(
            self,
            Kind::Op(Operator::Star)
                | Kind::Op(Operator::Slash)
                | Kind::Op(Operator::Plus)
                | Kind::Op(Operator::Minus)
                | Kind::Op(Operator::Greater)
                | Kind::Op(Operator::Less)
                | Kind::Op(Operator::LessEqual)
                | Kind::Op(Operator::GreaterEqual)
        )
    }

    pub fn identifier_from(value: &'tok str) -> Kind<'tok> {
        match value {
            "var" => Kind::Var,
            "const" => Kind::Const,
            "match" => Kind::Match,
            "if" => Kind::If,
            "else" => Kind::Else,
            "fun" => Kind::Fun,
            "struct" => Kind::Struct,
            "enum" => Kind::Enum,
            "return" => Kind::Return,
            _ => Kind::Value(Value::Ident(value)),
        }
    }
}

impl<'tok> IntoToken<'tok> for Kind<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(self, (start_byte, end_byte).into())
    }
}
