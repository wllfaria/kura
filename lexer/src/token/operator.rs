use std::fmt;

use super::{
    kind::Kind,
    token::{IntoToken, Token},
};

#[derive(Debug)]
pub enum Operator {
    LeftParen,
    RightParen,
    LeftBracket,
    RightBracket,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    MinusEqual,
    Plus,
    PlusEqual,
    Equal,
    ThickArrow,
    EqualEqual,
    Star,
    StarEqual,
    Ampersand,
    Slash,
    SlashEqual,
    Colon,
    SemiColon,
    Bang,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    NotEqual,
    And,
    Or,
}

impl fmt::Display for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Operator::Minus => write!(f, "-"),
            Operator::MinusEqual => write!(f, "-="),
            Operator::Colon => write!(f, ":"),
            Operator::SemiColon => write!(f, ";"),
            Operator::LeftParen => write!(f, "("),
            Operator::RightParen => write!(f, ")"),
            Operator::LeftBracket => write!(f, "["),
            Operator::RightBracket => write!(f, "]"),
            Operator::LeftBrace => write!(f, "{{"),
            Operator::RightBrace => write!(f, "}}"),
            Operator::Comma => write!(f, ","),
            Operator::Dot => write!(f, "."),
            Operator::Plus => write!(f, "+"),
            Operator::PlusEqual => write!(f, "+="),
            Operator::Equal => write!(f, "="),
            Operator::ThickArrow => write!(f, "=>"),
            Operator::EqualEqual => write!(f, "=="),
            Operator::Star => write!(f, "*"),
            Operator::StarEqual => write!(f, "*="),
            Operator::Ampersand => write!(f, "&"),
            Operator::Slash => write!(f, "/"),
            Operator::Bang => write!(f, "!"),
            Operator::Less => write!(f, "<"),
            Operator::LessEqual => write!(f, "<="),
            Operator::Greater => write!(f, ">"),
            Operator::GreaterEqual => write!(f, ">="),
            Operator::NotEqual => write!(f, "!="),
            Operator::SlashEqual => write!(f, "/="),
            Operator::And => write!(f, "&&"),
            Operator::Or => write!(f, "||"),
        }
    }
}

impl<'tok> IntoToken<'tok> for Operator {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(Kind::Op(self), (start_byte, end_byte).into())
    }
}
