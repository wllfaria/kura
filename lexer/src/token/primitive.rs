use core::fmt;

use super::{
    kind::Kind,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug)]
pub enum FloatSizes {
    F8,
    F16,
    F32,
    F64,
}

impl fmt::Display for FloatSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FloatSizes::F8 => write!(f, "f8"),
            FloatSizes::F16 => write!(f, "f16"),
            FloatSizes::F32 => write!(f, "f32"),
            FloatSizes::F64 => write!(f, "f64"),
        }
    }
}

#[derive(Debug)]
pub enum IntSizes {
    I8,
    I16,
    I32,
    I64,
    Isize,
}

impl fmt::Display for IntSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            IntSizes::I8 => write!(f, "i8"),
            IntSizes::I16 => write!(f, "i16"),
            IntSizes::I32 => write!(f, "i32"),
            IntSizes::I64 => write!(f, "i64"),
            IntSizes::Isize => write!(f, "isize"),
        }
    }
}

#[derive(Debug)]
pub enum UIntSizes {
    U8,
    U16,
    U32,
    U64,
    Usize,
}

impl fmt::Display for UIntSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            UIntSizes::U8 => write!(f, "u8"),
            UIntSizes::U16 => write!(f, "u16"),
            UIntSizes::U32 => write!(f, "u32"),
            UIntSizes::U64 => write!(f, "u64"),
            UIntSizes::Usize => write!(f, "usize"),
        }
    }
}

impl TryFrom<&str> for IntSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(IntSizes::I8),
            "i16" => Ok(IntSizes::I16),
            "i32" => Ok(IntSizes::I32),
            "i64" => Ok(IntSizes::I64),
            "isize" => Ok(IntSizes::Isize),
            _ => Err(()),
        }
    }
}

impl TryFrom<&str> for FloatSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "f8" => Ok(FloatSizes::F8),
            "f16" => Ok(FloatSizes::F16),
            "f32" => Ok(FloatSizes::F32),
            "f64" => Ok(FloatSizes::F64),
            _ => Err(()),
        }
    }
}

impl TryFrom<&str> for UIntSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "u8" => Ok(UIntSizes::U8),
            "u16" => Ok(UIntSizes::U16),
            "u32" => Ok(UIntSizes::U32),
            "u64" => Ok(UIntSizes::U64),
            "usize" => Ok(UIntSizes::Usize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Primitive {
    Bool(bool),
    UInt {
        value: u64,
        size: Option<UIntSizes>,
    },
    Int {
        value: i64,
        size: Option<IntSizes>,
    },
    Float {
        value: f64,
        size: Option<FloatSizes>,
    },
}

impl fmt::Display for Primitive {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Primitive::Bool(b) => write!(f, "{b}"),
            Primitive::UInt { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Primitive::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Primitive::Int { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
        }
    }
}

impl<'tok> IntoToken<'tok> for Primitive {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(
            Kind::Value(Value::Primitive(self)),
            (start_byte, end_byte).into(),
        )
    }
}
