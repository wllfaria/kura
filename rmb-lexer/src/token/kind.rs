use miette::Error;

use super::{
    operator::Operator,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug, PartialEq)]
pub enum Kind<'tok> {
    Value(Value<'tok>),
    Op(Operator),

    Var,
    Const,
    Match,
    If,
    Else,
    Fun,
    Struct,
    Enum,
    Return,
    Eof,
}

impl std::fmt::Display for Kind<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Kind::Value(val) => write!(f, "{}", val),
            Kind::Op(op) => write!(f, "{}", op),
            Kind::Var => write!(f, "var"),
            Kind::Const => write!(f, "const"),
            Kind::Match => write!(f, "match"),
            Kind::If => write!(f, "if"),
            Kind::Else => write!(f, "else"),
            Kind::Fun => write!(f, "fun"),
            Kind::Struct => write!(f, "struct"),
            Kind::Enum => write!(f, "enum"),
            Kind::Return => write!(f, "return"),
            Kind::Eof => write!(f, "eof"),
        }
    }
}

impl<'tok> Kind<'tok> {
    pub fn infix_precedence(&self) -> Result<(u8, u8), Error> {
        match self {
            Kind::Op(Operator::Plus) | Kind::Op(Operator::Minus) => Ok((1, 2)),
            Kind::Op(Operator::Less)
            | Kind::Op(Operator::LessEqual)
            | Kind::Op(Operator::EqualEqual)
            | Kind::Op(Operator::Or)
            | Kind::Op(Operator::Greater)
            | Kind::Op(Operator::GreaterEqual) => Ok((3, 4)),
            Kind::Op(Operator::Star) | Kind::Op(Operator::Slash) => Ok((5, 6)),
            _ => miette::bail!("invalid operator with no precedence"),
        }
    }

    pub fn is_binary_op(&self) -> bool {
        matches!(
            self,
            Kind::Op(Operator::Star)
                | Kind::Op(Operator::Slash)
                | Kind::Op(Operator::And)
                | Kind::Op(Operator::Or)
                | Kind::Op(Operator::Plus)
                | Kind::Op(Operator::Minus)
                | Kind::Op(Operator::Greater)
                | Kind::Op(Operator::EqualEqual)
                | Kind::Op(Operator::Less)
                | Kind::Op(Operator::LessEqual)
                | Kind::Op(Operator::NotEqual)
                | Kind::Op(Operator::GreaterEqual)
        )
    }

    pub fn identifier_from(value: &'tok str) -> Kind<'tok> {
        match value {
            "var" => Kind::Var,
            "const" => Kind::Const,
            "match" => Kind::Match,
            "if" => Kind::If,
            "else" => Kind::Else,
            "fun" => Kind::Fun,
            "struct" => Kind::Struct,
            "enum" => Kind::Enum,
            "return" => Kind::Return,
            _ => Kind::Value(Value::Ident(value)),
        }
    }
}

impl<'tok> IntoToken<'tok> for Kind<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(self, (start_byte, end_byte).into())
    }
}
