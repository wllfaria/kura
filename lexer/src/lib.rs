use miette::{Error, LabeledSpan};

mod token;

use token::{NumSize, Token};

struct Lexer<'lex> {
    pos: usize,
    source: &'lex str,
    complete_source: &'lex str,
}

impl<'lex> Lexer<'lex> {
    pub fn new(source: &'lex str) -> Self {
        Self {
            pos: 0,
            source,
            complete_source: source,
        }
    }
}

impl<'lex> Lexer<'lex> {
    fn lex_numerals(&mut self) -> Result<Token<'lex>, Error> {
        let end_of_numeral = self
            .source
            .find(|c| !matches!(c, '-' | '_' | '.' | '0'..='9'))
            .unwrap_or(self.source.len());

        let mut literal = &self.source[..end_of_numeral];

        let mut dotted = literal.splitn(3, '.');
        match (dotted.next(), dotted.next(), dotted.next()) {
            (Some(one), Some(two), Some(_)) => {
                // numeric literal contains two or more dots, so we only lex
                // from the first numerical to the last numerical before the
                // second dot, <one.len()>.<two.len()>
                //
                // 123.456.789 -> 123.456 (leave .789) to be lexed after
                literal = &literal[..one.len() + 1 + two.len()];
            }
            _ => {
                // literal is either a integer, or a valid floating pointer
                // numeral, so we don't have to do anything here
            }
        }

        let mut dashed = literal.splitn(3, '-');
        match (dashed.next(), dashed.next(), dashed.next()) {
            (Some(""), Some(two), Some(_)) => {
                // numeric literal contain two or more dashes, but one of the
                // dashes is the first character, meaning we should only lex the
                // middle part as a signed integer
                //
                // -123.456-123 ("", "123.456", "123") (result of splitn)
                literal = &literal[..two.len() + 1]; // +1 to account for the `-`
            }
            (Some(one), Some(_), Some(_)) => {
                // numeric literal has two or more dashes, and don't start with a
                // dash, so we only parse the first part and ignore the rest.
                //
                // 123-456.78-9 ("123", "456.78", "9")
                literal = &literal[..one.len()];
            }
            _ => {
                // numeral is a unsigned integer, so we don't have to do
                // anything here
            }
        }

        // NOTE: we have to update the lexer state here before we potentially
        // change the size of the literal after trimming underscores
        let bytes_eaten = literal.len();
        self.pos += bytes_eaten;
        self.source = &self.source[bytes_eaten..];

        // we accept underscores in numerals, but rust parser don't, so we
        // trim here before parsing to rust's parser
        let literal = literal.replace('_', "");

        // numbers are allowed to potentially have a postfix specifying its size
        // like `1234usize`
        let end_of_postfix = self
            .source
            .find(|c: char| c.is_whitespace())
            .unwrap_or(self.source.len());

        let mut postfix = &self.source[..end_of_postfix];
        if postfix.ends_with(';') {
            postfix = postfix.trim_matches(';');
        }

        let size = match NumSize::try_from(postfix) {
            Ok(size) => Some(size),
            Err(_) => None,
        };

        if size.is_some() {
            self.source = &self.source[postfix.len()..];
            self.pos += postfix.len();
        }

        let is_signed = literal.contains('-');
        let is_float = literal.contains('.');
        match (is_float, is_signed) {
            (false, true) => Ok(Token::SignedInteger {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(e) => return Err(miette::miette! {
                        labels = vec![
                            LabeledSpan::at(self.pos - bytes_eaten..self.pos, "this numeric literal"),
                        ],
                        "{e}",
                    }.with_source_code(self.complete_source.to_string())),
                },
                size,
            }),
            (false, false) => Ok(Token::Integer {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(e) => return Err(miette::miette! {
                        labels = vec![
                            LabeledSpan::at(self.pos - bytes_eaten..self.pos, "this numeric literal"),
                        ],
                        "{e}",
                    }.with_source_code(self.complete_source.to_string())),
                },
                size,
            }),
            (true, _) => Ok(Token::Float {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(e) => return Err(miette::miette! {
                        labels = vec![
                            LabeledSpan::at(self.pos - bytes_eaten..self.pos, "this numeric literal"),
                        ],
                        "{e}",
                    }.with_source_code(self.complete_source.to_string())),
                },
                size,
            }),
        }
    }

    fn advance_by(&mut self, amount: usize) {
        self.source = &self.source[amount..];
        self.pos += amount;
    }
}

impl<'lex> Iterator for Lexer<'lex> {
    type Item = Result<Token<'lex>, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let mut chars = self.source.chars().peekable();
            let c = chars.next()?;

            let mut simple_token = |token: Token<'lex>| {
                self.advance_by(c.len_utf8());
                Some(Ok(token))
            };

            let mut token_from_pairs = |pairs: Vec<(char, Token<'lex>)>, default: Token<'lex>| {
                let Some(next) = chars.peek() else {
                    return (default, 1);
                };

                for pair in pairs {
                    if next == &pair.0 {
                        return (pair.1, 2);
                    }
                }

                (default, 1)
            };

            match c {
                c if c.is_whitespace() => {
                    self.advance_by(c.len_utf8());
                    continue;
                }
                c if c.is_numeric() => return Some(self.lex_numerals()),
                '(' => return simple_token(Token::LeftParen),
                ')' => return simple_token(Token::RightParen),
                '[' => return simple_token(Token::LeftBracket),
                ']' => return simple_token(Token::RightBracket),
                '{' => return simple_token(Token::LeftBrace),
                '}' => return simple_token(Token::RightBrace),
                ':' => return simple_token(Token::Colon),
                ';' => return simple_token(Token::SemiColon),
                ',' => return simple_token(Token::Comma),
                '.' => return simple_token(Token::Dot),
                '+' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::AddAssign)], Token::Plus);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '=' => {
                    let (token, advance) = token_from_pairs(
                        vec![('=', Token::EqualEqual), ('>', Token::ThickArrow)],
                        Token::Equal,
                    );
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '*' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::MultiplyAssign)], Token::Star);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '&' => return simple_token(Token::Ampersand),
                '/' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::DivideAssign)], Token::Slash);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '!' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::NotEqual)], Token::Bang);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '<' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::LessEqual)], Token::Less);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '>' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::GreaterEqual)], Token::Greater);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '-' => match chars.peek() {
                    Some(c) if c.is_numeric() => return Some(self.lex_numerals()),
                    Some('=') => {
                        self.source = &self.source[c.len_utf8() + 1..];
                        self.pos += 2;
                        return Some(Ok(Token::MinusAssign));
                    }
                    _ => return simple_token(Token::Minus),
                },
                _ => {
                    let next_whitespace = self
                        .source
                        .find(|c| !matches!(c, 'a'..='z' | 'A'..='Z' | '_' | '0'..='9'))
                        .unwrap_or(self.source.len());

                    let identifier = &self.source[..next_whitespace];

                    self.advance_by(next_whitespace);

                    return Some(Ok(Token::identifier_from(identifier)));
                }
            };
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sut(source: &str) -> Lexer<'_> {
        Lexer::new(source)
    }

    #[test]
    fn lexing_numericals() {
        let source = [
            // positive integers and floats
            "12346.67890",
            "1234567890",
            "9__876___5.43____210",
            "9_876_5.43_210",
            "9_____876__543____210",
            "9_876_543_210",
            // negative integers and floats
            "-9_____876__5.43____210",
            "-123_46.678_90",
            "-12346.67890",
            "-1234567890",
            "-9_____876__543____210",
            "-9_876_543_210",
            "-9_876_-5-43_210",
            // integers and floats with postfixes
            "-123456789i32",
            "123456789usize",
            "123.456789f64",
            "123.456789f64",
            "-12f8",
            "3.14159265358979323846264338327950288_f32",
            "3.141592653f32;",
        ];
        let source = source.join("\n");

        let mut numerals = vec![];

        for token in make_sut(&source) {
            numerals.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(numerals);
    }

    #[test]
    fn lexing_punctuations() {
        let source = [
            "()", "[]", "{}", ",", ".", "+", "-", "=", "*", "&", "*=", "+=", "-=", "/=", "!", "!=",
            "==", "<=", ">=", "<", ">", "/", ":", ";",
        ];
        let source = source.join(" ");

        let mut punctuations = vec![];
        for token in make_sut(&source) {
            punctuations.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(punctuations);
    }

    #[test]
    fn lexing_builtin_identifiers() {
        let source = [
            "var", "const", "match", "if", "else", "fun", "struct", "enum", "return",
        ];

        let source = source.join(" ");

        let mut builtin_identifiers = vec![];
        for token in make_sut(&source) {
            builtin_identifiers.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(builtin_identifiers);
    }

    #[test]
    fn lexing_simple_codeblock() {
        let source = [
            "fun calculate_circumference(diameter: f64) => f64 {",
            "   const pi = 3.14159265358979323846264338327950288_f32;",
            "   const radius = diameter / 2.0;",
            "   const circumference = 2.0 * pi * radius;",
            "",
            "   return circumference;",
            "}",
        ];

        let source = source.join("\n");

        let mut calculate_circumference_function = vec![];
        for token in make_sut(&source) {
            calculate_circumference_function.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(calculate_circumference_function);
    }
}
