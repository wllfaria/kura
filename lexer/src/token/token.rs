use std::{
    fmt,
    ops::{Range, RangeBounds},
};

use miette::SourceSpan;

use super::kind::Kind;

pub trait IntoToken<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok>;
}

#[derive(Debug, Clone, Copy)]
pub struct Location {
    pub start_byte: usize,
    pub end_byte: usize,
}

impl Location {
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            start_byte: start,
            end_byte: end,
        }
    }
}

impl From<Location> for SourceSpan {
    fn from(value: Location) -> Self {
        SourceSpan::new(value.start_byte.into(), value.end_byte - value.start_byte)
    }
}

impl From<(usize, usize)> for Location {
    fn from((start_byte, end_byte): (usize, usize)) -> Self {
        Self {
            start_byte,
            end_byte,
        }
    }
}

impl From<Range<usize>> for Location {
    fn from(range: Range<usize>) -> Self {
        let start = match range.start_bound() {
            std::ops::Bound::Included(start) => *start,
            std::ops::Bound::Excluded(start) => *start,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        let end = match range.end_bound() {
            std::ops::Bound::Included(end) => *end,
            std::ops::Bound::Excluded(end) => *end,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        Location {
            start_byte: start,
            end_byte: end,
        }
    }
}

#[derive(Debug)]
pub struct Token<'tok> {
    pub kind: Kind<'tok>,
    pub location: Location,
}

impl<'tok> Token<'tok> {
    pub fn new(kind: Kind<'tok>, location: Location) -> Self {
        Self { kind, location }
    }
}

impl fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.kind)
    }
}
