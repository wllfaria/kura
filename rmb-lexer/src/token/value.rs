use std::fmt;

use super::primitive::Primitive;

#[derive(Debug, PartialEq)]
pub enum Value<'tok> {
    Primitive(Primitive),
    Ident(&'tok str),
    String(&'tok str),
}

impl fmt::Display for Value<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Value::Primitive(p) => write!(f, "{}", p),
            Value::Ident(i) => write!(f, "{i}"),
            Value::String(s) => write!(f, "{s}"),
        }
    }
}
