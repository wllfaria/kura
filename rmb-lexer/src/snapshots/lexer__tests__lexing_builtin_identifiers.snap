---
source: lexer/src/lib.rs
expression: builtin_identifiers
---
[
    Token {
        kind: Var,
        location: Location {
            start_byte: 0,
            end_byte: 3,
        },
    },
    Token {
        kind: Const,
        location: Location {
            start_byte: 4,
            end_byte: 9,
        },
    },
    Token {
        kind: Match,
        location: Location {
            start_byte: 10,
            end_byte: 15,
        },
    },
    Token {
        kind: If,
        location: Location {
            start_byte: 16,
            end_byte: 18,
        },
    },
    Token {
        kind: Else,
        location: Location {
            start_byte: 19,
            end_byte: 23,
        },
    },
    Token {
        kind: Fun,
        location: Location {
            start_byte: 24,
            end_byte: 27,
        },
    },
    Token {
        kind: Struct,
        location: Location {
            start_byte: 28,
            end_byte: 34,
        },
    },
    Token {
        kind: Enum,
        location: Location {
            start_byte: 35,
            end_byte: 39,
        },
    },
    Token {
        kind: Return,
        location: Location {
            start_byte: 40,
            end_byte: 46,
        },
    },
]
